ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerTestContext.scala

```diff
@@ -243,6 +328,23 @@ private[testtool] final class LedgerTestContext private[infrastructure] (
   ): Future[TransactionTree] =
     submitAndWaitRequest(party, exercise(party).command).flatMap(submitAndWaitForTransactionTree)
 
+  def exerciseAndGetContract[T](
+      party: Party,
+      exercise: Party => Primitive.Update[_]
+  ): Future[Primitive.ContractId[T]] =
+    submitAndWaitRequest(party, exercise(party).command)
+      .flatMap(submitAndWaitForTransaction)
+      .map(extractContracts)
+      .map(_.head.asInstanceOf[Primitive.ContractId[T]])
```

Comment:
Without this we would get all contracts created in the transaction instead of a single one and we would not be able to dictate what kind of contract it is. But I do agree that the casting is quite a wart. Not sure of how I could feed the type information correctly without that.

Reply:
Can we even make the assumption that the first contract is of type `T`? How about we pass the template companion object (`TemplateCompanion[T]`) or template identifier as parameter to the method, which which we can actually check the created events for matching template ids.

Reply:
See [this other comment](https://github.com/digital-asset/daml/pull/2713#discussion_r319888452) for more details.